File: src/main/java/com/abrastat/general/Species.java
package com.abrastat.general;

import java.io.*;
import javax.json.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


abstract class Species {

    private static final Logger LOGGER = LoggerFactory.getLogger(Species.class);

    //Json mapper variables
    private String species;
    private String[] allowedAbilities;
    private int baseHp, baseAttack, baseDefense, baseSpecialAttack, baseSpecialDefense, baseSpeed;
    private double height, weight;
    private Type[] types = new Type[2];
    private double genderRatio;

    protected Species(String speciesName)  {

        String pokedex = "./src/main/resources/pokedex.json";
        JsonObject jsonObject = null;

        try {
            JsonReader reader = Json.createReader(new FileInputStream(pokedex));
            jsonObject = reader.readObject();

        } catch (FileNotFoundException e) {
            LOGGER.error("Error reading Pokedex file at: {}", pokedex);
        }

        JsonObject jsonPokemon = jsonObject.getJsonObject(speciesName);
        this.setSpecies(speciesName);

        JsonObject jsonAbilities = jsonPokemon.getJsonObject("abilities");
        JsonObject jsonBaseStats = jsonPokemon.getJsonObject("baseStats");
        JsonArray jsonTypesArray = jsonPokemon.getJsonArray("types");

    //  this.setAbilities(jsonAbilities);
        this.setBaseStats(jsonBaseStats);
        this.setHeight(jsonPokemon);
        this.setWeight(jsonPokemon);
        this.setTypes(jsonTypesArray);

    }

    public String getSpecies()  {
        return this.species;
    }

    public void setSpecies(String speciesName)    {
        this.species = speciesName.substring(0, 1).toUpperCase() + speciesName.substring(1).toLowerCase();
    }

    public String[] getAllowedAbilities() {
        return allowedAbilities;
    }

    public void setAllowedAbilities(JsonObject allowedAbilities) {
        this.allowedAbilities[0] = allowedAbilities.getValue("0").toString();
        this.allowedAbilities[1] = allowedAbilities.getValue("1").toString();
        this.allowedAbilities[2] = allowedAbilities.getValue("H").toString();
    }

    public void setBaseStats(JsonObject baseStats) {
        this.baseHp =               baseStats.getInt("hp");
        this.baseAttack =           baseStats.getInt("attack");
        this.baseDefense =          baseStats.getInt("defense");
        this.baseSpecialAttack =    baseStats.getInt("special_attack");
        this.baseSpecialDefense =   baseStats.getInt("special_defense");
        this.baseSpeed =            baseStats.getInt("speed");
    }

    public int getBaseHp() {
        return baseHp;
    }

    public int getBaseAttack() {
        return baseAttack;
    }

    public int getBaseDefense() {
        return baseDefense;
    }

    public int getBaseSpecialAttack() {
        return baseSpecialAttack;
    }

    public int getBaseSpecialDefense() {
        return baseSpecialDefense;
    }

    public int getBaseSpeed() {
        return baseSpeed;
    }

    public double getHeight() {
        return height;
    }

    public void setHeight(JsonObject jsonHeight) {
        height = jsonHeight.getJsonNumber("height").doubleValue();
    }

    public double getWeight() {
        return weight;
    }

    public void setWeight(JsonObject weight) {
        this.weight = weight.getJsonNumber("weight").doubleValue();
    }

    public Type[] getTypes() {
        return types;
    }

    public void setTypes(JsonArray types) {
        final String type0 = types.getJsonString(0)
                .toString()
                .replace("\"", "")
                .toUpperCase();
        this.types[0] = Type.valueOf(type0);

        try {
            final String type1 = types.getJsonString(1)
                    .toString()
                    .replace("\"", "")
                    .toUpperCase();
            this.types[1] = Type.valueOf(type1);

        } catch (ArrayIndexOutOfBoundsException e)  {
            System.out.println("no secondary typing for " + getSpecies() + ".");
            this.types[1] = Type.NONE;
        }
    }

}

File: src/main/java/com/abrastat/general/Species.kt
package com.abrastat.general

import org.slf4j.LoggerFactory
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.util.*
import javax.json.Json
import javax.json.JsonArray
import javax.json.JsonObject

abstract class Species protected constructor(speciesName: String) {
    //Json mapper variables
    var species: String? = null
        private set
    val allowedAbilities = arrayOfNulls<String>(3)
    var baseHp = 0
        private set
    var baseAttack = 0
        private set
    var baseDefense = 0
        private set
    var baseSpecialAttack = 0
        private set
    var baseSpecialDefense = 0
        private set
    var baseSpeed = 0
        private set
    var baseSpecial = 0
        private set
    var height = 0.0
        private set
    var weight = 0.0
        private set
    @JvmField
    val types = arrayOfNulls<Type>(2)
    private val genderRatio = 0.0

    init {
        val pokedex = "./src/main/resources/pokedex.json"
        var jsonObject: JsonObject? = null
        try {
            val reader = Json.createReader(FileInputStream(pokedex))
            jsonObject = reader.readObject()
        } catch (e: FileNotFoundException) {
            LOGGER.error("Error reading Pokedex file at: {}", pokedex)
        }
        val jsonPokemon = jsonObject!!.getJsonObject(speciesName)
        setSpecies(speciesName)
        val jsonAbilities = jsonPokemon.getJsonObject("abilities")
        val jsonBaseStats = jsonPokemon.getJsonObject("baseStats")
        val jsonTypesArray = jsonPokemon.getJsonArray("types")

        //  this.setAbilities(jsonAbilities);
        setBaseStats(jsonBaseStats)
        setHeight(jsonPokemon)
        setWeight(jsonPokemon)
        setTypes(jsonTypesArray)
    }

    private fun setSpecies(speciesName: String) {
        species = speciesName.substring(0, 1).uppercase(Locale.getDefault()) + speciesName.substring(1).lowercase(Locale.getDefault())
    }

    fun setAllowedAbilities(allowedAbilities: JsonObject) {
        this.allowedAbilities[0] = allowedAbilities.getValue("0").toString()
        this.allowedAbilities[1] = allowedAbilities.getValue("1").toString()
        this.allowedAbilities[2] = allowedAbilities.getValue("H").toString()
    }

    private fun setBaseStats(baseStats: JsonObject) {
        baseHp = baseStats.getInt("hp")
        baseAttack = baseStats.getInt("attack")
        baseDefense = baseStats.getInt("defense")
        baseSpecialAttack = baseStats.getInt("special_attack")
        baseSpecialDefense = baseStats.getInt("special_defense")
        baseSpeed = baseStats.getInt("speed")
        baseSpecial = if (baseStats.isNull("special")) 0 else
            when (baseStats.getString("special")) {
                "special_attack" -> baseSpecialAttack
                "special_defense" -> baseSpecialDefense
                else -> 0
            }
    }

    private fun setHeight(jsonHeight: JsonObject) {
        height = jsonHeight.getJsonNumber("height").doubleValue()
    }

    private fun setWeight(weight: JsonObject) {
        this.weight = weight.getJsonNumber("weight").doubleValue()
    }

    private fun setTypes(types: JsonArray) {
        val type0 = types.getJsonString(0)
                .toString()
                .replace("\"", "")
                .uppercase(Locale.getDefault())
        this.types[0] = Type.valueOf(type0)
        try {
            val type1 = types.getJsonString(1)
                    .toString()
                    .replace("\"", "")
                    .uppercase(Locale.getDefault())
            this.types[1] = Type.valueOf(type1)
        } catch (e: Exception) {
            println("no secondary typing for $species.")
            this.types[1] = Type.NONE
        }
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(Species::class.java)
    }
}